File: stacks/validate-stack-sequences/validate-stack-sequences.java
class Solution {
  public boolean validateStackSequences(int[] pushed, int[] popped) {
    int n = pushed.length;

    if (n == 0) {
      return true;
    }

    int i = 0;

    Stack<Integer> stack = new Stack<>();

    for (int p : pushed) {
      stack.push(p);

      while (!stack.isEmpty() && stack.peek() == popped[i]) {
        stack.pop();
        i++;
      }
    }
    return i == n;
  }
}

File: stacks/validate-stack-sequences/validate-stack-sequences.ts
function validateStackSequences(pushed: number[], popped: number[]): boolean {
  const n: number = pushed.length;

  if (n === 0) {
    return true;
  }

  let i: number = 0;
  const stack: number[] = [];

  for (const p of pushed) {
    stack.push(p);

    while (stack.length && stack[stack.length - 1] === popped[i]) {
      stack.pop();
      i++;
    }
  }
  return i === n;
}
